Fixing CurrentReplicas and CurrentRevision in completeRollingUpdate

adilGhaffarDev opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
In `completeRollingUpdate` function we should not compare `status.UpdatedReplicas` and `status.ReadyReplicas` with `status.Replicas` because `status.Replicas `is not showing total desired replicas, it shows only those replicas that are created, but here we want to compare with total desired replicas(set.Spec.Replicas) and only when `set.Spec.Replicas == status.ReadyReplicas` and `set.Spec.Replicas == status.UpdatedReplicas` we should change `status.CurrentReplicas` to `status.UpdatedReplicas`. This was causing updated image on pods with ordinal number lower than the rolling partition number when they were being deleted.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/119685 https://github.com/kubernetes/kubernetes/issues/119684

#### Special notes for your reviewer:
This is taking e2e test from this PR https://github.com/kubernetes/kubernetes/pull/119759 , lot of discussion regarding the fix happened there. Just for reference.
#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed the issue where pod with ordinal number lower than the rolling partitioning number was being deleted it was coming up with updated image.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```
Thanks @liangyuanpeng for adding e2e tests for this. Integration is added by @mimowo , in addition, @mimowo  also helped in fixing e2e test.
Co-authored-by: Lan Liang <gcslyp@gmail.com>. Michał Woźniak @mimowo

adilGhaffarDev committed:
Fixing CurrentReplicas and CurrentRevision in completeRollingUpdate

lentzi90 commented:
/ok-to-test

aleksandra-malinowska commented:
Thank you Adil for picking this up!

I believe this is also going to fix #120700.

/lgtm

aleksandra-malinowska commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -587,8 +587,8 @@ func inconsistentStatus(set *apps.StatefulSet, status *apps.StatefulSetStatus) b
-		status.UpdatedReplicas == status.Replicas &&
-		status.ReadyReplicas == status.Replicas {
+		status.UpdatedReplicas == *set.Spec.Replicas &&
+		status.ReadyReplicas == *set.Spec.Replicas {
```
Nit: I'm wondering if we might also need a check for `status.Replicas == *set.Spec.Replicas` and/or `status.CurrentReplicas == 0`? 

I don't _think_ we can get here now if we have condemned pods that haven't been successfully removed, but given the complexity of this controller's logic and the strong guarantee we need from this check (to avoid unexpected rollout progress), I think it might make it more resilient to check all conditions we care about.

adilGhaffarDev replied:
I agree, I will add this check too

adilGhaffarDev replied:
added that check kindly check

liangyuanpeng commented:
Have some lint of e2e test and i can help to fix it at tomorrow.

adilGhaffarDev commented:
> Have some lint failed of e2e test and i can help to fix it at tomorrow.

No worries, It was minor fix, I did it myself. Thanks.

adilGhaffarDev commented:
@aojea and @mimowo kindly check it and if it looks good can we merge this soon.

mimowo commented:
@adilGhaffarDev please update the release note

aleksandra-malinowska commented:
/lgtm

mimowo commented:
Are you going to cherry-pick this for 1.27 and 1.28? It seems required to cherry-pick https://github.com/kubernetes/kubernetes/pull/120398.

adilGhaffarDev commented:
> Are you going to cherry-pick this for 1.27 and 1.28? It seems required to cherry-pick #120398.

yes we need to cherry-pick, should I wait for merge just do it right away?

mimowo commented:
/assign @soltysh

aojea commented:
I still don't understand what is the the root cause, checking git blame this code is the same for the last 6 years,

why the test started to flake https://github.com/kubernetes/kubernetes/issues/120700#issuecomment-1723179950 after https://github.com/kubernetes/kubernetes/pull/120398 ?

why does this say it fixes https://github.com/kubernetes/kubernetes/issues/119684 and that bug only happens in 1.27.1?

adilGhaffarDev commented:
> I still don't understand what is the the root cause, checking git blame this code is the same for the last 6 years,

Yes that's been there for a long time, but I believe its wrong because `status.replicas` contains created replicas not desired replicas and here we need to check with desired replicas. And why it stopped working recently, is because of recent addition of  `isCreated` check here: https://github.com/kubernetes/kubernetes/pull/117865/files?diff=split&w=0#diff-a1e25eb50e4b2aed8947c2caa5d6d1ddfef1a5f2f95fdf71f2388c657823ee13R324 which was not there before and we were incrementing status.replicas without checking if the pod was created or not. Without `isCreated` check we get status.replicas equal to Spec.replicas.

> why the test started to flake https://github.com/kubernetes/kubernetes/issues/120700#issuecomment-1723179950 after https://github.com/kubernetes/kubernetes/pull/120398 ?

@aleksandra-malinowska can explain more about this flake.

> why does this say it fixes https://github.com/kubernetes/kubernetes/issues/119684 and that bug only happens in 1.27.1? what has changed in 1.27.1?

I have tried the reproduction steps provided by the reporter here : https://github.com/kubernetes/kubernetes/issues/119684#issuecomment-1660127079 in 1.28.0 too and issue is there too. Issue is in all versions that have this PR: https://github.com/kubernetes/kubernetes/pull/117865 or has isCreated check when incrementing status.replicas.

liggitt commented:
Before making changes, and _especially_ before backporting them, we need to understand when the regression happened (what minor and what patch release) by seeing where the test fails. When we can see a working minor/patch and a broken minor/patch, then we can try to understand what changes went into the statefulset controller in that range and see how to restore previous functionality.

Without clearly understanding this, we're at risk for introducing new regressions trying to fix old regressions.

liggitt commented:
Collecting the relevant merges to master that changed statefulset behavior, along with the patch versions they were backported to, would help in understanding what we've regressed and how to recover

aleksandra-malinowska commented:
Timeline of changes and their impact:
- #115331 introduced a bug which caused StatefulSets to not recreate a pod completed successfully (bug: #118310). If the pods in test exited with code 0, they weren't recreated. This probably prevented the detection of current bug (recreating at wrong revision, which was introduced later) in e2e tests.
- #117865 introduced another bug: `status.Replicas` is updated with a delay - it's showing too small value immediately after the pods are created (not counting them), but self-correcting the next time the object is processed (bug: #120732). Semantics of `status.Replicas` remained unchanged: all existing (created and not deleted) pods.
- Both of the above changes have already been backported some months ago. I'm not sure if rolling them back now is a viable option.
- While the test was passing (not detecting the second issue), a consequence of miscalculated status was reported in #119684 and #119685: we rely on this value to complete rolling update. The first attempt to fix it (removing `IsCreated() check)  *would change the semantics of `status.Replicas`*, overestimating the number of replicas: #119759. 
- #120398 fixed the issue with not recreating a pod. This caused the test to fail (bug: #120700) and detect the issue with pods recreated at wrong revision. We can roll it back and merge again later, if getting tests to green is a priority - I believe this will just conceal the bug with wrong revision again.
- This PR (#120731, in case this is referenced elsewhere) is the second attempt to fix the issue with wrong revision: it doesn't use `status.Replicas` to determine rolling update progress, but rather `spec.Replicas`. I believe this is more resilient, since it's a perfectly valid for `status.Replica` to be lower or higher than desired replicas temporarily. If we ever reached complete rolling update check in such state, we'd be dealing with the same issue again. I recognise a lot of StatefulSet controller logic is based on such invariants, but I believe it's safer to check the actual conditions needed to safely complete rollback.
- We should definitely also fix `status.Replicas` calculation (bug: #120732), however we don't yet have a PR with this fix.
- Finally, I believe the failing test is likely accepting a pod in `Succeeded` phase as having the correct revision. (This is speculation as I haven't digged into this test - @mimowo might have more context). If so, this test would be more robust if it only considered running pods.

@mimowo @adilGhaffarDev please correct me if I missed something.

liggitt commented:
@aleksandra-malinowska that's super helpful, thanks... can you annotate which releases (and patch-releases, if backported) those changes went into?

adilGhaffarDev commented:
> * This PR (

Yes I agree with you, this PR is making the check more resilient, with this PR we are relying on spec.replicas instead of status.replicas, because status.replicas are not reliable as of right now (@aleksandra-malinowska  has already opened a issue for this https://github.com/kubernetes/kubernetes/issues/120732). Also when reliability of status.replicas is fixed, I think we should still compare with spec.replicas in completeRollingUpdate, I believe its the right way to do it, we want to check if all replica pods that statefulset can create are updated and status.replicas does not give that info, it gives created replicas at that moment, they might be less than the total desired replicas. So I think we should consider this PR even if we decide to revert other PRs, because this might cause issues in future. Another thing is we are also adding e2e test for this specific issue https://github.com/kubernetes/kubernetes/issues/119685 to make sure this fix works.

aleksandra-malinowska commented:
>can you annotate which releases (and patch-releases, if backported) those changes went into?

Added it to above summary, reposting for visibility:
- #115331 versions: 1.27.0
- #117865 versions: 1.28.0, 1.27.4, 1.26.7, 1.25.12

#115331 was in fact not cherry-picked, but made it in time for 1.27.0.

aleksandra-malinowska commented:
@liggitt I created revert PR for #120755 (most recent change which likely exposed the bug in tests) to get dashboard green again.

soltysh reviewed (commented):
/hold

Based on the discussion I agree we should revert to a "working" state to better understand the root cause. 

I personally tested this on 1.26.6, 1.27.4 and 1.28.2 (which I have quick access to) and it seems like it's not a easy thing to reproduce, since it won't happen always. I agree that we should figure out the right approach especially why https://github.com/kubernetes/kubernetes/pull/120398 is triggering that failure, maybe the root cause is in that PR, which I haven't reviewed.

adilGhaffarDev commented:
> I personally tested this on 1.26.6, 1.27.4 and 1.28.2 (which I have quick access to) and it seems like it's not a easy thing to reproduce, since it won't happen always. I agree that we should figure out the right approach especially why #120398 is triggering that failure, maybe the root cause is in that PR, which I haven't reviewed.

If you are talking about the unexpected images on pods with an ordinal number less than the partitioning number(https://github.com/kubernetes/kubernetes/issues/119685), I have been able to reproduce it every time with the steps that I have mentioned here: https://github.com/kubernetes/kubernetes/pull/119759#issuecomment-1711813950
I tested on 1.26.6, 1.27.3 and 1.28.0, on 1.26.6 I could not reproduce it and on 1.27.3 and 1.28.0 it was reproducible all the time.

aleksandra-malinowska commented:
>I tested on 1.26.6, 1.27.3 and 1.28.0, on 1.26.6 I could not reproduce it and on 1.27.3 and 1.28.0 it was reproducible all the time.

If it's reproducible on 1.27.3 and 1.27.1 (as reported in #119684),  it would have been introduced before #117865 (cherry-picked to 1.27.4). Are there any versions on which it _doesn't_ happen?

Given that this PR almost certainly fixes it (as per @mimowo's experiments - 40 runs without flakes), I suspect the original bug might have been related to miscalculating replica count as well.

adilGhaffarDev commented:
> If it's reproducible on 1.27.3 and 1.27.1 (as reported in #119684), it would have been introduced before #117865 (cherry-picked to 1.27.4). Are there known 1.27 versions (like 1.27.0?) on which it _doesn't_ happen?

I checked 1.27.0 and I can reproduce the issue there too every time. So this issue has been there since the release of 1.27.0.

adilGhaffarDev commented:
@soltysh I am removing the fix from this PR, so we can see e2e failures.

adilGhaffarDev commented:
test failed:

```
 I0920 08:10:25.517]   [FAILED] Current revision should not equal update revision during rolling update
I0920 08:10:25.517]   Expected
I0920 08:10:25.517]       <string>: ss2-5459d8585b
I0920 08:10:25.517]   not to equal
I0920 08:10:25.517]       <string>: ss2-5459d8585b
I0920 08:10:25.517]   In [It] at: test/e2e/apps/statefulset.go:576 @ 09/20/23 08:10:04.357 
```
This is the new test that we are adding, to test this scenario: https://github.com/kubernetes/kubernetes/issues/119684#issuecomment-1660127079

soltysh commented:
/test pull-kubernetes-e2e-gce

liggitt commented:
is it expected the e2e test would have passed in this PR without the fix merged?

liggitt commented:
> * [Parallel StatefulSet pod create & delete #117865](https://github.com/kubernetes/kubernetes/pull/117865) versions: 1.28.0, 1.27.4, 1.26.7, 1.25.12

If that PR was a performance optimization, should we do a straight rollback in 1.25+ to remove the bug it introduced?

adilGhaffarDev commented:
> is it expected the e2e test would have passed in this PR without the fix present in the PR?

seems like it will not fail 100% of the time. As @soltysh shared here: https://kubernetes.slack.com/archives/C18NZM5K9/p1695235870995709?thread_ts=1695032177.473889&cid=C18NZM5K9 

> is there a more deterministic test (unit or integration test) that would reproduce the bug more consistently?
I will try to find a scenario where it fails 100%, if there is one I will add the test for it.

aleksandra-malinowska commented:
>The top priority is reverting regressions that got introduced in patch releases (https://github.com/kubernetes/kubernetes/pull/117865)

We are no longer convinced this PR introduced the bug. The issue is reproducible in 1.27.0, which doesn't have it.

I'm trying to understand it better as the hypothesis "we are just miscalculating newly created pods" just got much weaker:
- the bug happens in some versions before the change to calculating status (1.27.0-1.27.3 confirmed)
- the bug is indeterministic. If the problem was due to creating pods and completing rolling update in the same iteration, I believe the bug would have been deterministic (at least unless an error occurs and ends the iteration prematurely).
- the "newly created pod not counted" scenario implies that the pod has already been created _before_ the complete upgrade check. This could cause the upgrade to complete prematurely, but the pod would have been created _before_ this completion, so it should still be at the current revision. Only deleting it a second time could trigger the creation at updated revision.

What we do know:
- changing rolling update check to use `set.spec.Replicas` fixes the issue -> the error happens due to comparison with `status.Replicas`
- counting all replicas (previous fix attempt) also prevents this issue, even though it changes the semantics of status -> we are probably not counting some pods in `status.Replicas`

The only situation where we don't count some pods is when they have empty pod phase. We've assumed so far that this can only happen if the pod was created in the same iteration. I'm going to run some more experiments to verify this.

aleksandra-malinowska commented:
So we may still complete rolling upgrade (and I believe we do) if we exit `updateStatefulSet` with error: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L118

When the StatefulSet observes a pod in `Failed` phase, it'll try to recreate it. On 1.26.0, this works immediately in a kind cluster. On 1.27.0, the pod is deleted successfully, but the create call fails for a while with `object is being deleted: pods "web-am-0" already exists, the server was not able to generate a unique name for the object`, then succeeds. 

On both versions after a single failed create call, we exit with error and a non-nil status with `Replicas=1` (we never increased the counter as the create failed). Because status isn't nil, we proceed to `updateStatefulSetStatus`, where we do the complete rolling upgrade check.

[Pod Lifecycle](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/) documentation suggests what may have triggered this to occur somewhat reliably in 1.27.0: pods linger in Failed/Succeeded phase before being deleted from the API server:
>Since Kubernetes 1.27, the kubelet transitions deleted Pods, except for [static Pods](https://kubernetes.io/docs/tasks/configure-pod-container/static-pod/) and [force-deleted Pods](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination-forced) without a finalizer, to a terminal phase (Failed or Succeeded depending on the exit statuses of the pod containers) before their deletion from the API server.

This also explains indeterminism, and particularly @soltysh's results with the bug occurring more rarely when there are 3 master nodes instead of 1 - it probably increases latency sufficiently for a pod to not be caught in "failed, but not yet really deleted state" as often.

So the bug is probably very long standing, and it's triggered by a failed create call for the pod 0 when partition=1 and other pods are already updated. That's _very_ specific and even if observed before 1.27, would have been hard to reproduce without simulating API server failure for this specific call. 

We can verify this hypothesis by trying to force delete the pod and checking if (as per documentation) it's deleted immediately and prevents the error from occurring.

aleksandra-malinowska commented:
(Actually, there's a possibility that this could have failed for more pods than pod 0, if all current replicas were deleted at once and the first create call failed. I don't think it changes much as it's even more unlikely to occur than for a single pod.)

Exiting with error and status values that correspond to reality, then updating the actual status based on it seems reasonable - we may have changed something at this point, although it probably wouldn't hurt to skip complete rolling upgrade check if we exit with an error. Checking `status.Replicas` instead of `set.Spec.Replicas` is the questionable part - why should we rely on this value, which may be lower or higher than desired replicas, is unclear to me. I agree it makes sens to still check it - to make sure there are no extra pods - but what we really need to be looking at here is desired replicas.

aleksandra-malinowska commented:
>We can verify this hypothesis by trying to force delete the pod and checking if (as per documentation) it's deleted immediately and prevents the error from occurring.

Just verified this on a 1.27.0 kind cluster - the issue doesn't occur if the pod is deleted with `--force --grace-period=0`. It occurs if it's deleted without them.

aleksandra-malinowska commented:
>The top priority is reverting regressions that got introduced in patch releases (#117865)

#117865 didn't make this bug worse in either 1.28.0 or 1.27.4.

It might have introduced it in 1.26.7 and 1.25.12, depending how you look at it (the issue was still there, but it was largely dormant).

Can we even roll back the cherry-picks without rolling back on the master? If yes, I'd say 1.26.7 revert is the only one that makes sense - 1.25 is no longer getting new patch releases AFAIK, and 1.27/1.28 won't be fixed by the rollback.

>The next priority is reverting regressions that got introduced in minor releases (#115331)

Similarly, #115331 didn't exactly introduce the bug in 1.27.0, but made it worse (exposed it).

>The next priority is fixing bugs if we can prove we're fixing them safely (e.g. they pass tests we add to catch the regressions that crept in)

This PR fixes a long-standing bug that became exposed in 1.27.0. As for the reliability of regression tests, perhaps adding a finalizer with several minutes' sleep can help with that. @adilGhaffarDev do you think we can try that?

Other than the principles mentioned above, does it matter how large are the changes? It would likely be much more complex in this case to rollback the somehow-related large PRs, some of which may require additional rollbacks (@mimowo can probably assess the impact better, as his change was part of a larger effort) or may not be possible as clean rollbacks. As opposed to this, the fix is really quite small and simple. 

I'd say the mass-rollback approach has significantly more potential for unintended consequences and introducing new bugs than the fix in this case.

adilGhaffarDev commented:
> This PR fixes a long-standing bug that became exposed in 1.27.0. As for the reliability of regression tests, perhaps adding a finalizer with several minutes' sleep can help with that. @adilGhaffarDev do you think we can try that?

@aleksandra-malinowska Thank you for your investigation. I will update the test by adding some DeletionGracePeriodSeconds.

lentzi90 commented:
> DeletionGracePeriodSeconds

The grace period will only do something if the workload is not exiting on its own. I believe the test workload is mostly just nginx so it will exit quickly and not "wait" for the grace period. You would need a workload that ignores `SIGTERM` or something like that. It would then get `SIGKILL` after the grace period.

adilGhaffarDev commented:
/test pull-kubernetes-e2e-gce

lentzi90 commented:
> @lentzi90 do you think this would work?
> 
> ```
> pod0name := getStatefulSetPodNameAtIndex(0, ss)
> 			pod0, err := c.CoreV1().Pods(ns).Get(ctx, pod0name, metav1.GetOptions{})
> 			e2epod.NewPodClient(f).Update(ctx, pod0name, func(pod *v1.Pod) {
> 				pod0.Spec.Containers[0].Lifecycle = &v1.Lifecycle{
> 					PreStop: &v1.LifecycleHandler{
> 						Exec: &v1.ExecAction{
> 							Command: []string{"/bin/sh", "-c", "sleep 500"},
> 						},
> 					},
> 				}
> 			})
> ```

I'm not sure. What I could find in the docs is [this](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks):

> the hook must complete before the TERM signal to stop the container can be sent. The Pod's termination grace period countdown begins before the PreStop hook is executed, so regardless of the outcome of the handler, the container will eventually terminate within the Pod's termination grace period.

But I'm not sure really how to interpret that. :shrug: Does it just skip sending `SIGTERM` if the PreStop hook is taking too long? It is worth a try at least :slightly_smiling_face:

liggitt commented:
> 1.25 is no longer getting new patch releases AFAIK

Just a note that 1.25 has ~1 more patch release before EOL, though risk/size of a revert would need to be considered

adilGhaffarDev commented:
@soltysh @aleksandra-malinowska @liggitt @mimowo can we have a consensus on the resolution of this issue? I can add it to the agenda of the next sig-apps community meeting and we can discuss it there too if that is needed.

liggitt commented:
> @soltysh @aleksandra-malinowska @liggitt @mimowo can we have a consensus on the resolution of this issue? I can add it to the agenda of the next sig-apps community meeting and we can discuss it there too if that is needed.

I'll defer to @soltysh for review of the tests and fix, but I would expect this to be accompanied by a unit or integration test with a 100% reproduce rate for the issue being fixed (that type of test can mock out pod termination behavior and should be able to trigger the condition that breaks better than a racy e2e test).

Use that test to show when the regression happened (the test alone, without this fix, should pass against some past minor release). We can then include the test and the fix in subsequent minor releases to resolve the regression.

mimowo commented:
> @lentzi90 do you think this would work?
> ```
> pod0name := getStatefulSetPodNameAtIndex(0, ss)
> 			pod0, err := c.CoreV1().Pods(ns).Get(ctx, pod0name, metav1.GetOptions{})
> 			e2epod.NewPodClient(f).Update(ctx, pod0name, func(pod *v1.Pod) {
> 				pod0.Spec.Containers[0].Lifecycle = &v1.Lifecycle{
> 					PreStop: &v1.LifecycleHandler{
> 						Exec: &v1.ExecAction{
> 							Command: []string{"/bin/sh", "-c", "sleep 500"},
> 						},
> 					},
> 				}
> 			})
> ```

I don't think this works. For one, you cannot modify the container: 
https://github.com/kubernetes/kubernetes/blob/26c3f66887d13efc8cd627b50beeeef060c2e39c/pkg/apis/core/validation/validation.go#L4936. I'm not sure why it does not fail here, but I don't see it updated when I fetch the pod afterwards. Also, this fails: 
```golang
			pod0.Spec.Containers[0].Lifecycle = &v1.Lifecycle{
				PreStop: &v1.LifecycleHandler{
					Exec: &v1.ExecAction{
						Command: []string{"/bin/sh", "-c", "sleep 500"},
					},
				},
			}
			/pod0, err = c.CoreV1().Pods(ns).Update(ctx, pod0, metav1.UpdateOptions{})
```

Second, I think the issue isn't with delaying the deletion, either with graceful termination, or finalizers. I think the issue occurs when the pod is deleted completely, so the `status.Replicas` is 2, as the field counts any existing pods: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L324-L326.

The current version of the test deletes the pod0 from the code test, and this is why it fails. However, it does not fail consistently, as the stateful set controller may recreate the pod before checking the `completeRollingUpdate` condition. So, the longer the pod is deleted and not recreated, the the test would fail (if not for the fix). However, I'm not quite sure  we can delay the pod recreation in e2e test to make it fail 100% times.

I would suggest keeping the e2e test, but also adding an integration test, Here we can be triggering the sts reconciliations on demand, to ensure `completeRollingUpdate` is called before the pod is recreated.

Finally, the issue with the `completeRollingUpdate` is coincidentally meaning that since https://github.com/kubernetes/kubernetes/pull/115331, we have a bug exposed in https://github.com/kubernetes/kubernetes/pull/120398. The same fix will work. However, given the complicated structure of dependencies between bugs I suggest we put effort to write down good tests, following the sentiment of https://github.com/kubernetes/kubernetes/pull/120731#issuecomment-1737323262. In particular, I would suggest that we also add an integration test as part of re-introducing https://github.com/kubernetes/kubernetes/pull/120398 (for the scenario there). @aleksandra-malinowska wdyt? I can help with the integration test.

aleksandra-malinowska commented:
>Second, I think the issue isn't with delaying the deletion, either with graceful termination, or finalizers. I think the issue occurs when the pod is deleted completely, so the status.Replicas is 2, as the field counts any existing pods: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L324-L326.

The issue also occurs in 1.27.0 (before the linked code was submitted).

What happens then:
- we count the pod as created: https://github.com/kubernetes/kubernetes/blob/v1.27.0/pkg/controller/statefulset/stateful_set_control.go#L317
- we decrease the number of replicas on deletion of failed pod: https://github.com/kubernetes/kubernetes/blob/v1.27.0/pkg/controller/statefulset/stateful_set_control.go#L414
- we replace the pod in `replicas` slice with the new one (without phase == not created): https://github.com/kubernetes/kubernetes/blob/v1.27.0/pkg/controller/statefulset/stateful_set_control.go#L416
- we fail to create a pod _and never increase replicas again_: https://github.com/kubernetes/kubernetes/blob/v1.27.0/pkg/controller/statefulset/stateful_set_control.go#L434

mimowo commented:
> The issue also occurs in 1.27.0 (before the linked code was submitted).

So IIUC it will also flake on 1.27.0. Probably worth checking experimentally, as well as for older versions (I will try).

I suppose it would also flake on older versions, since the `completeRollingUpdate` looks broken for a long time: https://github.com/kubernetes/kubernetes/blame/master/pkg/controller/statefulset/stateful_set_utils.go#L579-L590. Probably from the beginning. It's just that it was hidden by the fact that when `restartPolicy: Always` then prior to https://github.com/kubernetes/kubernetes/pull/115331 getting `Failed` phase was rare (but possible in a couple scenarios, especially with finalizers). Getting `Succeeded` phase was impossible with core k8s (unless set by external component).

Another option is that if we have an integration test failing consistently we may be iterate faster to find out how old versions are afffected.

adilGhaffarDev commented:
@mimowo I have added an integration test and also updated my e2e test to align with your changes. I have not been able to create an exact scenario in the integration test, kindly take a look and let me know if we can fix it. Also, check the updated e2e test. 
I am changing the title to wip.

adilGhaffarDev commented:
e2e tests failed where we wanted them to fail:

-  https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/120731/pull-kubernetes-e2e-kind/1712075194014633984
- https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/120731/pull-kubernetes-e2e-kind-ipv6/1712075194064965632

I will fix the type check fix in the integration test and run the tests again.

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +510,152 @@ var _ = SIGDescribe("StatefulSet", func() {
+			framework.ExpectNoError(err)
+			hostLabel := "kubernetes.io/hostname"
+			hostLabelVal := readyNode.Labels[hostLabel]
+			ss.Spec.Template.Spec.NodeSelector = map[string]string{hostLabel: hostLabelVal}
```
Is this needed? If no, I would suggest to simplify.

adilGhaffarDev replied:
It is not needed. The reproduction rate of this issue was lower when there were multiple nodes so I thought forcing pods to schedule on the same node might make the test fail more consistently, but it didn't work, I will remove it.

adilGhaffarDev replied:
its removed as suggested

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +510,152 @@ var _ = SIGDescribe("StatefulSet", func() {
+			pod0, err = c.CoreV1().Pods(ss.Namespace).Update(ctx, pod0, metav1.UpdateOptions{})
+			framework.ExpectNoError(err)
+			pods.Items[0] = *pod0
+			defer func() {
```
Try to avoid complicated defer, also you can delete the finalizer with the utility function:
`defer e2epod.NewPodClient(f).RemoveFinalizer(ctx, pod0.Name, testFinalizer)`

adilGhaffarDev replied:
changed it as suggested

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +510,152 @@ var _ = SIGDescribe("StatefulSet", func() {
+			pod0name := getStatefulSetPodNameAtIndex(0, ss)
+			pod0, err := c.CoreV1().Pods(ns).Get(ctx, pod0name, metav1.GetOptions{})
+			framework.ExpectNoError(err)
+			pod0.Finalizers = []string{"example.com/myfinalizer"}
```
It's probably better to append, rather than set.

adilGhaffarDev replied:
changed it as suggested

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +510,152 @@ var _ = SIGDescribe("StatefulSet", func() {
+					currentRevision))
+			}
+
+			ginkgo.By("ADDING FINALIZER to pod0")
```
```suggestion
			ginkgo.By("Adding finalizer for pod0 to induce an error or pod recreation")
```

adilGhaffarDev replied:
changed it as suggested

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +510,152 @@ var _ = SIGDescribe("StatefulSet", func() {
+					_, err = c.CoreV1().Pods(ss.Namespace).Update(ctx, pod0, metav1.UpdateOptions{})
+					if err == nil {
+						ginkgo.By("REMOVING FINALIZER to POD0 - success")
+						pods = e2estatefulset.GetPodList(ctx, c, ss)
```
I would suggest to do asserts in the main body of the test, rather than defer

adilGhaffarDev replied:
changed it as suggested

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +510,152 @@ var _ = SIGDescribe("StatefulSet", func() {
+				})
+				framework.ExpectNoError(err)
+			}()
+			// Delete all of the old pod.
```
```suggestion
			ginkgo.By("Delete the old pod0 so that it is not counted in status")
```

adilGhaffarDev replied:
changed it as suggested

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +510,152 @@ var _ = SIGDescribe("StatefulSet", func() {
+			currentRevision, updateRevision = ss.Status.CurrentRevision, ss.Status.UpdateRevision
+			gomega.Expect(currentRevision).ToNot(gomega.Equal(updateRevision), "Current revision should not equal update revision during rolling update")
+
+			pods = e2estatefulset.GetPodList(ctx, c, ss)
```
ginkgo.By("Verify pod images")

adilGhaffarDev replied:
changed it with a more explanatory sentence.

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +510,152 @@ var _ = SIGDescribe("StatefulSet", func() {
+
+			ginkgo.By(fmt.Sprintf("Updating stateful set template: update image from %s to %s", oldImage, newImage))
+			gomega.Expect(oldImage).ToNot(gomega.Equal(newImage), "Incorrect test setup: should update to a different image")
+			ss, err = updateStatefulSetWithRetries(ctx, c, ns, ss.Name, func(update *appsv1.StatefulSet) {
```
I think we should add the finalizer before here. From now on the update is rolling, so potentially we may be adding the finalizer too late. It probably does not matter in practice though.

adilGhaffarDev replied:
changed it as suggested

mimowo commented on test/integration/statefulset/statefulset_test.go:
```diff
@@ -489,6 +489,137 @@ func TestAutodeleteOwnerRefs(t *testing.T) {
 	}
 }
 
+func TestDeletingPodForRollingUpdatePartition(t *testing.T) {
```
Maybe we could drop it given the e2e test can reproduce it consistently. WDYT @soltysh ?

adilGhaffarDev replied:
Updated the test according to your PR: https://github.com/kubernetes/kubernetes/pull/120921 , now integration should always fail.

mimowo replied:
Please resolve, I support going with e2e + integration. 
- point for e2e: it demonstrates the issue with kubelet, which was changed in 1.27+ that every pod transitions to terminal phase before deletion.
- point for integration: it demonstates the issue with handling of Failed phase, which is very tricky to reproduce before 1.27. It requires eviction, for example due to node pressure.

mimowo commented on test/integration/statefulset/statefulset_test.go:
```diff
@@ -489,6 +489,138 @@ func TestAutodeleteOwnerRefs(t *testing.T) {
+		sts.Spec.Template.Spec.Containers[0].Image = newImage
+	})
+
+	// Await for the pods 1 and 2 to be recreated, while pod0 remains running
```
```suggestion
	// Await for the pod1 to be recreated, while pod0 remains running
```

adilGhaffarDev replied:
fixed it as suggested

mimowo commented on test/integration/statefulset/statefulset_test.go:
```diff
@@ -489,6 +489,138 @@ func TestAutodeleteOwnerRefs(t *testing.T) {
+				recreatedPods.Items = append(recreatedPods.Items, pod)
+			}
+		}
+		setPodsReadyCondition(t, c, &v1.PodList{Items: recreatedPods.Items}, v1.ConditionTrue, time.Now().Add(-120*time.Minute))
```
TBH, I copy-pasted this `.Add(-120*time.Minute))`. Do we need it? If not, please simplify.

adilGhaffarDev replied:
I have removed it.

mimowo commented on test/integration/statefulset/statefulset_test.go:
```diff
@@ -489,6 +489,138 @@ func TestAutodeleteOwnerRefs(t *testing.T) {
+	}
+	// Setting all pods in Running, Ready, and Available
+	// by setting LastTransitionTime to more than 3600 seconds ago
+	setPodsReadyCondition(t, c, &v1.PodList{Items: pods.Items}, v1.ConditionTrue, time.Now().Add(-120*time.Minute))
```
TBH, I copy-pasted this `.Add(-120*time.Minute))`. Do we need it? If not, please simplify.

adilGhaffarDev replied:
removed

mimowo commented on test/integration/statefulset/statefulset_test.go:
```diff
@@ -489,6 +489,138 @@ func TestAutodeleteOwnerRefs(t *testing.T) {
+		}
+		return ss.Status.ReadyReplicas == *ss.Spec.Replicas-1, nil
+	}); err != nil {
+		t.Fatalf("failed to verify .Spec.Template.Spec.Containers[0].Image is updated for sts %s: %v", sts.Name, err)
```
```suggestion
		t.Fatalf("failed to await for pod0 to be not counted as ready is status of sts %s: %v", sts.Name, err)
```

adilGhaffarDev replied:
changed it as suggested

mimowo commented on test/integration/statefulset/statefulset_test.go:
```diff
@@ -489,6 +489,138 @@ func TestAutodeleteOwnerRefs(t *testing.T) {
+		setPodsReadyCondition(t, c, &v1.PodList{Items: recreatedPods.Items}, v1.ConditionTrue, time.Now().Add(-120*time.Minute))
+		return ss.Status.UpdatedReplicas == *ss.Spec.Replicas-*sts.Spec.UpdateStrategy.RollingUpdate.Partition && ss.Status.Replicas == *ss.Spec.Replicas && ss.Status.ReadyReplicas == *ss.Spec.Replicas, nil
+	}); err != nil {
+		t.Fatalf("failed to verify .Spec.Template.Spec.Containers[0].Image is updated for sts %s: %v", sts.Name, err)
```
```suggestion
		t.Fatalf("failed await for pod1 to be recreated by sts %s: %v", sts.Name, err)
```

adilGhaffarDev replied:
changed it as suggested

mimowo commented on test/integration/statefulset/statefulset_test.go:
```diff
@@ -489,6 +489,138 @@ func TestAutodeleteOwnerRefs(t *testing.T) {
+		setPodsReadyCondition(t, c, &v1.PodList{Items: recreatedPods.Items}, v1.ConditionTrue, time.Now().Add(-120*time.Minute))
+		return len(recreatedPods.Items) > 0, nil
+	}); err != nil {
+		t.Fatalf("failed to verify .Spec.Template.Spec.Containers[0].Image is updated for sts %s: %v", sts.Name, err)
```
```suggestion
		t.Fatalf("failed await for pod0 to be recreated by sts %s: %v", sts.Name, err)
```

adilGhaffarDev replied:
changed it as suggested

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +512,155 @@ var _ = SIGDescribe("StatefulSet", func() {
+				}
+			}
+
+			ginkgo.By("Marking pod0 as Failed so that StatefulSet controller recreates it")
```
I would suggest to just delete the pod (as below), and drop this section. Deleting the pod corresponds better to the scenario described in the issues. It also makes the test code agnostic to which phase is used by kubelet underneath. 

The section was used by me to demonstrate it was a problem before 1.27.

Yes, this will reproduce the issue only 1.27+. However, let's keep it simple until we have a decision to cherry-pick beyond. 

If we decide to cherry-pick beyond, we may consider a couple of options, including dropping the test and relying on integration test, or adjusting the test to set the phase manually.

adilGhaffarDev replied:
Ok, I have removed it.

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +512,155 @@ var _ = SIGDescribe("StatefulSet", func() {
 
 		})
 
+		ginkgo.It("should perform canary updates and phased rolling updates of template modifications for partiton1 and delete pod-0", func(ctx context.Context) {
```
```suggestion
		ginkgo.It("should perform canary updates and phased rolling updates of template modifications for partiton1 and delete pod0", func(ctx context.Context) {
```
nit: alternatively, use pod-0 throughout the test consistently.

adilGhaffarDev replied:
changed it to pod-0 throughout the test

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +512,155 @@ var _ = SIGDescribe("StatefulSet", func() {
+			pods.Items[0] = *pod0
+			defer e2epod.NewPodClient(f).RemoveFinalizer(ctx, pod0.Name, testFinalizer)
+
+			ginkgo.By("Updatinf image on StatefulSet")
```
```suggestion
			ginkgo.By("Updating image on StatefulSet")
```

adilGhaffarDev replied:
fixed

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +512,155 @@ var _ = SIGDescribe("StatefulSet", func() {
+			})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Delete the old pod0 so that it is not counted in status")
```
```suggestion
			ginkgo.By("Deleting the pod so that kubelet terminates it and StatefulSet controller recreates it")
```
Note that, it is not deleted immediately after calling `deleteStatefulPodAtIndex` due to the finalizer.

adilGhaffarDev replied:
changed it as suggested

adilGhaffarDev commented on test/integration/statefulset/util.go:
```diff
@@ -139,6 +140,75 @@ func newSTS(name, namespace string, replicas int) *appsv1.StatefulSet {
 }
 
+// newSTS returns a StatefulSet with a fake container image and RollingUpdate partitionNumber
+func newSTSWithPartitionNumber(name, namespace string, replicas int, partitionNumber int) *appsv1.StatefulSet {
```
removed it as suggested.

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +512,145 @@ var _ = SIGDescribe("StatefulSet", func() {
 
 		})
 
+		ginkgo.It("should perform canary updates and phased rolling updates of template modifications for partiton1 and delete pod-0", func(ctx context.Context) {
```
IIUC the test in the current version passes on master because the deleted pod-0 ends in Succeeded phase, and this is not recreated by StatefulSet controller because https://github.com/kubernetes/kubernetes/pull/120398 got reverted. 

Because testing both scenarios makes sense (one reproduces now, the other will prevent regression related to the other PR), I suggest to make two versions of the test, one with basic ss, the other with extra failing container. For example: 
```golang
	ss.Spec.Template.Spec.Containers = append(ss.Spec.Template.Spec.Containers, v1.Container{
		Name:    "sleep-exit-with-1",
		Image:   imageutils.GetE2EImage(imageutils.BusyBox),
		Command: []string{"sh", "-c"},
		Args: []string{`
                                  _term(){
                                     echo "Received SIGTERM signal"
                                     exit 1
                                  }
                                  trap _term SIGTERM
                                  while true; do
                                     echo "Running..."
                                     sleep 1
                                  done
			`,
		},
	})
```

One way to commonize the code test it to wrap the test body in a function (parametrized with ss and potentially other stuff), similarly as it is done with the use of `rollbackTest`.

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +512,44 @@ var _ = SIGDescribe("StatefulSet", func() {
+			ss := e2estatefulset.NewStatefulSet("ss2", ns, headlessSvcName, 3, nil, nil, labels)
+			deletingPodForRollingUpdatePartitionTest(ctx, c, ns, ss)
+
+			ginkgo.By("Creating a new StatefulSet with failing container")
```
I would suggest to use dedicated `ginkgo.It` for the test case to make sure the test cases are truly independent.

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +512,44 @@ var _ = SIGDescribe("StatefulSet", func() {
+				Name:    "sleep-exit-with-1",
+				Image:   imageutils.GetE2EImage(imageutils.BusyBox),
+				Command: []string{"sh", "-c"},
+				Args: []string{`if [ "${POD_INDEX##*-}" = "0" ]; then
```
can we simplify the if by matching the entire pod name? This should be deterministic.

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +512,46 @@ var _ = SIGDescribe("StatefulSet", func() {
+				Name:    "sleep-exit-with-1",
+				Image:   imageutils.GetE2EImage(imageutils.BusyBox),
+				Command: []string{"sh", "-c"},
+				Args: []string{`if [ "${POD_NAME}" = "0" ]; then
```
Just checking, isn't the POD_NAME = `ss3-0` in this case?

adilGhaffarDev replied:
yes I missed that, let me push again,

mimowo replied:
Thanks, feel free to mark resolved.

mimowo reviewed (commented):
The tests LGTM, once you have a prove that they on Github, please put back the fix, in a separate commit and remove the "wip" from the PR title.

Vyom-Yadav commented:
Hey @adilGhaffarDev
Just curious whether this is planned for 1.29 or not. If it is, we should add the milestone label, which would help the release team track this issue.

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +512,46 @@ var _ = SIGDescribe("StatefulSet", func() {
+									_term(){
+										echo "Received SIGTERM signal"
+										exit 1
+									}
```
Maybe we could do the `if` inside the trap to just exit with 0 or 1 depending on the pod name. I think in the current form the container for pod-1 and pod-2 will continuously get restarted as it exits with 0 (since restartPolicy=Always).

adilGhaffarDev replied:
I have fixed it now it should be fine. Thank you for the help.

mimowo replied:
Thanks, feel free to mark resolved.

adilGhaffarDev commented:
> Hey @adilGhaffarDev
> Just curious whether this is planned for 1.29 or not. If it is, we should add the milestone label, which would help the release team track this issue.

Yes I believe we want it in 1.29, @soltysh @mimowo can you confirm?

mimowo commented:
> Yes I believe we want it in 1.29, @soltysh @mimowo can you confirm?

1.29 would be great, especially as it unblocks a fix for another issue which is anticipated, see [here](https://github.com/kubernetes/kubernetes/issues/118310#issuecomment-1761923920)

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +512,51 @@ var _ = SIGDescribe("StatefulSet", func() {
+		ginkgo.It("should perform canary updates and phased rolling updates of template modifications for partiton1 and delete pod-0 with failing container", func(ctx context.Context) {
+			ginkgo.By("Creating a new StatefulSet with failing container")
+			ss1 := e2estatefulset.NewStatefulSet("ss3", ns, headlessSvcName, 3, nil, nil, labels)
+			ss1.Spec.Template.Spec.Containers = append(ss.Spec.Template.Spec.Containers, v1.Container{
```
```suggestion
			ss.Spec.Template.Spec.Containers = append(ss.Spec.Template.Spec.Containers, v1.Container{
```
Or `ss1`, but it needs to be consistent throughout the test. I think `ss` it preferred to avoid confusion like here.

adilGhaffarDev replied:
sorry I missed that, now it's fixed.

mimowo replied:
Thanks, feel free to mark resolved.

mimowo reviewed (commented):
/lgtm
/assign @soltysh

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -510,6 +510,99 @@ var _ = SIGDescribe("StatefulSet", func() {
 
 		})
 
+		framework.ConformanceIt("should perform canary updates and phased rolling updates of template modifications for partiton1 and delete pod-0", func(ctx context.Context) {
```
Please resolve

Vyom-Yadav commented:
Excellent, I will add it to the 1.29 milestone.

/milestone v1.29

soltysh reviewed (approved):
There's still issue of the mismatching statuses, as I explained in slack, but I'll also post here for visibility:

While testing this PR I found STS status isn't also fully correct, I started logging out full status right after each wait command and I'm seeing this:
```
// CurrentRevision is the only revision, and the CurrentReplicas matches Replicas, ie. 3
STEP: 1: Status v1.StatefulSetStatus{ObservedGeneration:1, Replicas:3, ReadyReplicas:3, CurrentReplicas:3, UpdatedReplicas:3, CurrentRevision:"ss2-7b6c9599d5", UpdateRevision:"ss2-7b6c9599d5", AvailableReplicas:3}
// after first update, with Partition=1, we need to update only 2 pods (1 and 2), but CurrentReplicas (ie. old) should be 0, UpdatedReplicas (ie. new) should be 2
STEP: 2: Status v1.StatefulSetStatus{ObservedGeneration:2, Replicas:3, ReadyReplicas:3, CurrentReplicas:2, UpdatedReplicas:0, CurrentRevision:"ss2-7b6c9599d5", UpdateRevision:"ss2-5459d8585b", AvailableReplicas:3}
// after the pod removal, we need to recreate the pod for CurrentRevision, so CurrentReplicas=1, and UpdatedReplicas=2
STEP: 3: Status v1.StatefulSetStatus{ObservedGeneration:2, Replicas:3, ReadyReplicas:3, CurrentReplicas:1, UpdatedReplicas:2, CurrentRevision:"ss2-7b6c9599d5", UpdateRevision:"ss2-5459d8585b", AvailableReplicas:3}
```
if one checks the descriptions we have in [our API](https://github.com/kubernetes/api/blob/b01b44926aa4920c8c8c008003e16316cf59ffda/apps/v1/types.go#L276-L296) we clearly are doing a bad job reflecting the actual changes happening to Current(Replicas|Revision) and Updated(Replicas|Revisions) such that they don't reflect the actual state of world, since the division between Current and Updated should be strictly tied to Partitions (as described in [the original proposal](https://github.com/kubernetes/design-proposals-archive/blob/main/apps/statefulset-update.md)).

Still this PR improves the current situation and fixes parts of the problem we're seeing, and adds sufficient tests allowing us to ensure that `.spec.partition` is working as expected.

/hold cancel
/lgtm
/approve

soltysh commented:
/triage accepted
/priority backlog

soltysh commented:
Also, based on the discussion @mimowo @adilGhaffarDev and I had on slack, we're going to backport this to only 1.28 and 1.27 due to limitations. More details can be found in [this slack thread](https://kubernetes.slack.com/archives/C18NZM5K9/p1695032177473889). Although, I'd like to see us fixing that status I described above before attempting the picks, since there's still time before the next patch releases.
